miniwin/miniwin/src/miniwin_ddraw.cpp

```diff
@@ -171,36 +171,63 @@ HRESULT DirectDrawImpl::EnumDisplayModes(
 	LPDDENUMMODESCALLBACK lpEnumModesCallback
 )
 {
-	DDSURFACEDESC ddsd = {};
-	ddsd.dwSize = sizeof(DDSURFACEDESC);
-	ddsd.dwFlags = DDSD_HEIGHT | DDSD_WIDTH | DDSD_PIXELFORMAT;
-	ddsd.dwWidth = 640;
-	ddsd.dwHeight = 480;
-	ddsd.lPitch = 0;
-
-	ddsd.ddpfPixelFormat.dwSize = sizeof(DDPIXELFORMAT);
-	ddsd.ddpfPixelFormat.dwFlags = DDPF_RGB | D3DDD_DEVICEZBUFFERBITDEPTH;
-	ddsd.ddpfPixelFormat.dwRGBBitCount = 16; // Game only accpets 8 or 16
-
-	if (!lpEnumModesCallback(&ddsd, lpContext)) {
+	if (!lpEnumModesCallback) {
+		return DDERR_INVALIDPARAMS;
+	}
+
+	SDL_DisplayID displayID = SDL_GetPrimaryDisplay();
+	if (!displayID) {
 		return DDERR_GENERIC;
 	}
 
-	return S_OK;
+	SDL_DisplayMode** modes = SDL_GetFullscreenDisplayModes(displayID, nullptr);
+	if (!modes) {
+		return DDERR_GENERIC;
+	}
+
+	HRESULT status = S_OK;
+	for (SDL_DisplayMode** mode = modes; *mode != nullptr; mode++) {
+		const SDL_PixelFormatDetails* format = SDL_GetPixelFormatDetails((*mode)->format);
+		if (!format) {
+			continue;
+		}
+
+		DDSURFACEDESC ddsd = {};
+		ddsd.dwSize = sizeof(DDSURFACEDESC);
+		ddsd.dwFlags = DDSD_WIDTH | DDSD_HEIGHT;
+		ddsd.dwWidth = (*mode)->w;
+		ddsd.dwHeight = (*mode)->h;
```

Comment:
I think this is easier to read.

```suggestion
	int count_modes;
	SDL_DisplayMode** modes = SDL_GetFullscreenDisplayModes(displayID, &count_modes);
	if (!modes) {
		return DDERR_GENERIC;
	}

	HRESULT status = S_OK;
	for (int i = 0; i < count_modes; i++) {
		const SDL_PixelFormatDetails* format = SDL_GetPixelFormatDetails(modes[i]->format);
		if (!format) {
			continue;
		}

		DDSURFACEDESC ddsd = {};
		ddsd.dwSize = sizeof(DDSURFACEDESC);
		ddsd.dwFlags = DDSD_WIDTH | DDSD_HEIGHT;
		ddsd.dwWidth = modes[i]->w;
		ddsd.dwHeight = modes[i]->h;
```